### PR TITLE
#13 status code

### LINKED ISSUE #13: Error 500 on an existing rule in cloudflare
1. Send request:

```
POST http://localhost:4346/api/bans
204
1080 ms
```
```
POST /api/bans HTTP/1.1
X-Analyzer-Id: X-Analyzer-Id header params!!!!!!
Content-Type: application/json
User-Agent: PostmanRuntime/7.29.0
Accept: */*
Postman-Token: 2349d6ab-907e-4720-b833-16a0c221a53a
Host: localhost:4346
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
Content-Length: 167
 ```

Request body:
```json
{
  "target": {
    "ip": "1.2.3.5",
    "user_agent": "curl user-agent"
  },
  "reason": "reasonreasonreasonreasonreasonreasonreason",
  "ttl": 3000
}
```
 **Response:**
HTTP/1.1 204 No Content
date: Wed, 22 Jun 2022 14:41:42 GMT

2. Repeat request.
response body:
```json
{
  "code": 500,
  "reason": "CF responded unsuccessfull: [\"filter at index (0) conflicts with an existing filter\"])",
  "details": null
}
```

### LINKED ISSUE #16: Filters are not removed when deleting rules in cloudflare
1. Send POST /bans (target) from exuter-cloudflare
**Result**: response code 204, creating a rule inside cloudflare + creating a filter inside cloudflare.
2. Send DELETE /bans (target) from exuter-cloudflare
**Result**: response code 204, deleting a rule inside cloudflare + NOT deleting a filter inside cloudflare.
3. Send POST /bans (target) from exuter-cloudflare
**Result**: response code 500, NOT creating a rule inside cloudflare becouse exist a filter inside cloudflare.
**Response body:**
```
{
    "code": 500,
    "reason": "CF responded unsuccessfull: ["filter at index (0) conflicts with an existing filter"])",
    "details": null
}
```
It is necessary to remove unused filters in order to re-create rules

### PR DESCRIPTION
**Pull request by citizensix**
Closes #13 
Closes #16

### CODE REVIEW

File: src/executor/op_run.rs [hunk 1]
```diff
@@ -48,36 +48,58 @@ impl Executor for ExecutorService {
...
-            .create_block_rule(firewall_rule.clone(), models::RestrictionType::Block)
-            .await
+        let rule_ids = schema::nongratas::table
+            .filter(schema::nongratas::restriction_value.eq(firewall_rule.clone()))
+            .select(schema::nongratas::rule_id)
+            .load::<String>(&*conn)
             .map_err(ServerError::from)?;
+        if rule_ids.len() > 0 {
+            let target = schema::nongratas::table
+                .filter(schema::nongratas::restriction_value.eq(firewall_rule.clone()));
```
Comment (stuffy-idler) [inline src/executor/op_run.rs:58]: Whether `firewall_rule` should be cloned? Can it be used by ref?

File: src/executor/op_run.rs [hunk 2]
```diff
@@ -48,36 +48,58 @@ impl Executor for ExecutorService {
         let firewall_rule =
             models::form_firewall_rule_expression(rule.target.ip, rule.target.user_agent)
                 .ok_or(errors::ServerError::MissingTarget)?;
-
-        let rule_id = self
-            .client
-            .create_block_rule(firewall_rule.clone(), models::RestrictionType::Block)
-            .await
+        let rule_ids = schema::nongratas::table
+            .filter(schema::nongratas::restriction_value.eq(firewall_rule.clone()))
```
Comment (stuffy-idler) [inline src/executor/op_run.rs:52]: It should be also filtered for requested `analyzer_id`. Moreover there can only one rule_id for pair `target + analyzer_id`

File: src/executor/op_run.rs [hunk 3]
```diff
@@ -48,36 +48,58 @@ impl Executor for ExecutorService {
...
+                .execute(&*conn)
+                .map_err(ServerError::from)?;
+        } else {
+            let rule_id = self
+                .client
+                .create_block_rule(firewall_rule.clone(), models::RestrictionType::Block)
+                .await
+                .map_err(ServerError::from)?;
+
+            if block_request.ttl == 0 {
```
Comment (stuffy-idler) [inline src/executor/op_run.rs:79]: I guess it should be moved to API `handler`

File: src/executor/op_run.rs [hunk 4]
```diff
@@ -48,36 +48,58 @@ impl Executor for ExecutorService {
         let firewall_rule =
```
Comment (stuffy-idler) [inline src/executor/op_run.rs:48]: At 39 line there is `block_request` cloning – it's unnecessary and can be dropped. `info` and another macros takes parameters by ref

**Review by stuffy-idler — CHANGES_REQUESTED** [review]

**Review by stuffy-idler — APPROVED** [review]

**Comment by citizensix** [comment]
> @stuffy-idler Don't work! For all requests **/bans** there is response:
> 
> ```
> {
>     "code": 500,
>     "reason": "NotFound",
>     "details": null
> }
> ```
The problem was that `first()` treats empty rows as an error unlike `load()` so I've changed this

**Comment by stuffy-idler** [comment]
@Rabid-Squirrel

**Review by stuffy-idler — APPROVED** [review]

**Review by stuffy-idler — APPROVED** [review]

**Comment by stuffy-idler** [comment]
@citizensix PR can be merged

**Pull request merged**